### PR TITLE
feat(core): apply new ref scheme for directories and files

### PR DESCRIPTION
**Pull request by grouville**
Follow-up of https://github.com/dagger/dagger/pull/7511 (rebased on top of it)

This applies the new ref scheme to directories and files (moving out from the buildkit syntax), whilst remaining retrocompatible

#### Implementation

This does not change the cmd/flags logic as we only convert the new ref scheme to buildkit's ref scheme whilst remaining retrocompatible

As part of the implementation, we move the parseRefString implementation to the vcs package, to avoid dependency hell issue between the core and the CLI.

We introduce a ParseRef Func type, aimed to be used by ConvertToBuildKitRef to handle the two specific use-cases:
parseRefStringDir and parseRefStringFile

We rename parseRefString to parseRefStringDir, as it was intended for directory ref parsing, and we duplicate its logic for files. The only difference is the stat type: parseRefStringFile makes sure that the file exists locally as part of the implicit local file interpolation

### CODE REVIEW

File: engine/vcs/convert_test.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+		{
+			name:           "Git ref with version",
+			urlStr:         "git@github.com:sipsma/daggerverse@v0.9.1",
+			expectedUrlStr: "ssh://git@github.com:sipsma/daggerverse#v0.9.1",
+		},
+		{
+			name:           "Git ref with .git and version",
+			urlStr:         "git@github.com:sipsma/daggerverse.git@v0.9.1",
+			expectedUrlStr: "ssh://git@github.com:sipsma/daggerverse.git#v0.9.1",
+		},
```
Comment (jedevc) [inline engine/vcs/convert_test.go:82]: Hm, this shouldn't be a valid syntax.

The supported syntaxes should either be:
- The older buildkit ref style
- The newer module ref style

I don't think this is either of the above?

Comment (grouville) [reply engine/vcs/convert_test.go:82]: Ok 👌 
I wanted to cover all the test cases of the current TestParseGit: https://github.com/dagger/dagger/blob/main/cmd/dagger/module_test.go#L148

Comment (jedevc) [reply engine/vcs/convert_test.go:82]: Oh agreed, it's just that this specific case isn't in there.

**Comment by grouville** [comment]
Failure seems unrelated to implementation

**Comment by gerhard** [comment]
Want to rebase @grouville as soon as we have upgraded our CI to `v0.11.8`? Pretty sure there will be a PR from @jedevc for this later today.

**Comment by grouville** [comment]
> Want to rebase @grouville as soon as we have upgraded our CI to `v0.11.8`? Pretty sure there will be a PR from @jedevc for this later today.

I think we should review / rebase it once #7708 gets merged as it will impose the ref style for private git repo

**Comment by jedevc** [comment]
x-linking https://github.com/dagger/dagger/issues/6957 - not entirely sure how these should interact.

File: cmd/dagger/flags.go [hunk 2]
```diff
@@ -218,25 +221,29 @@ func (v *directoryValue) Get(ctx context.Context, dag *dagger.Client, modSrc *da
...
-			KeepGitDir: true,
-		}
-		if authSock, ok := os.LookupEnv("SSH_AUTH_SOCK"); ok {
-			gitOpts.SSHAuthSocket = dag.Host().UnixSocket(authSock)
-		}
-		gitDir := dag.Git(parsedGit.Remote, gitOpts).Branch(parsedGit.Fragment.Ref).Tree()
-		if subdir := parsedGit.Fragment.Subdir; subdir != "" {
-			gitDir = gitDir.Directory(subdir)
+	bk := &osBuildkitClient{}
+	ref, kind := vcs.ConvertToBuildKitRef(ctx, v.String(), bk, vcs.ParseRefStringDir)
```
Comment (jedevc) [inline cmd/dagger/flags.go:225]: We should be able to remove this function. Instead, we have `parseGit` call `vcs.ParseRef`, and then we can extract host, subpath, etc from there :tada:

We should avoid the dance of converting a parsed string into a raw string to parse it again - that's going to more error prone and difficult to follow :scream:

**Pull request closed without merging**